### PR TITLE
Adds eval based generator

### PR DESCRIPTION
**Pull request by aaronjeline**
Adds a generator that generates arbitrary types and arbitrary expressions of that type.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+#[derive(Debug, Clone)]
+struct FuzzTargetInput {
+    /// generated schema
+    pub schema: Schema,
+    /// generated entity slice
+    pub entities: Entities,
+    /// top level type of the policy
+    pub toplevel_type: SchemaType,
+    /// generated policy
+    pub expression: Expr,
```
Comment (cdisselkoen) [inline cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs:47]: the comments refer to a "policy", but it's really an expression right?

File: cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+        ];
+        let all_entities = Entities::try_from(hierarchy).map_err(|_| Error::NotEnoughData)?;
```
Comment (cdisselkoen) [inline cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs:90]: why is this the right error

File: cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+        let requests = [
+            schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
+            // schema.arbitrary_request(&hierarchy, u)?,
```
Comment (cdisselkoen) [inline cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs:88]: nit: commented-out code

File: cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    let mut path = PathBuf::new();
+    path.push("/tmp");
+    path.push("pols");
+    path.push(name);
+    let mut f = std::fs::File::create(path).unwrap();
+    write!(f, "{p}").unwrap();
+}
+
+// The main fuzz target. This is for type-directed fuzzing of ABAC
+// hierarchy/policy/requests
```
Comment (cdisselkoen) [inline cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs:165]: comment needs updating?

File: cedar-drt/fuzz/src/lib.rs [hunk 5]
```diff
@@ -15,15 +15,17 @@
  */
 
 mod dump;
+use cedar_policy_core::evaluator::{EvaluationError, Evaluator};
+use cedar_policy_core::extensions::Extensions;
```
Comment (cdisselkoen) [inline cedar-drt/fuzz/src/lib.rs:19]: can we put these next to the other imports from `cedar_policy_core`?

File: cedar-drt/src/lib.rs [hunk 6]
```diff
@@ -102,6 +117,76 @@ impl<'j> DefinitionalEngine<'j> {
...
+            expr,
+            expected,
+        })
+        .expect("Failed to serialize request");
+        self.thread
+            .new_string(request)
+            .expect("failed to create Java object for eval request string")
+    }
+
+    fn deserialize_eval_reponse(&self, response: JValue) -> bool {
```
Comment (cdisselkoen) [inline cedar-drt/src/lib.rs:164]:
```suggestion
    fn deserialize_eval_response(&self, response: JValue) -> bool {
```

File: cedar-drt/fuzz/src/lib.rs (hunk 5 above)
Comment (cdisselkoen) [reply cedar-drt/fuzz/src/lib.rs:19]: bump? right here they're between `mod dump` and `pub use dump::*` which is confusing

File: cedar-policy-generators/src/err.rs [hunk 7]
```diff
@@ -46,6 +46,8 @@ pub enum Error {
         /// short string describing what we were doing when this happened
         doing_what: &'static str,
     },
+    /// An error that occurs attempting to convert a hierarchy into an entities structure
+    EntitiesError(String),
```
Comment (cdisselkoen) [inline cedar-policy-generators/src/err.rs:50]: where do we actually construct this variant?

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy-generators/src/err.rs (hunk 7 above)
Comment (aaronjeline) [reply cedar-policy-generators/src/err.rs:50]: Converting a hierarchy into an `Entities` struct in `eval-type-directed`

**Review by cdisselkoen — APPROVED** [review]

File: cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+/// An ABAC hierarchy, policy, and 8 associated requests
+#[derive(Debug, Clone)]
+struct FuzzTargetInput {
+    /// generated schema
+    pub schema: Schema,
+    /// generated entity slice
+    pub entities: Entities,
+    /// generated expression
+    pub expression: Expr,
+    /// the requests to try for this hirarchy and policy. We try 8 requests per
```
Comment (khieta) [inline cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs:43]:
```suggestion
    /// the requests to try for this hierarchy and policy. We try 8 requests per
```

File: cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+            Schema::arbitrary_request_size_hint(depth),
+            Schema::arbitrary_request_size_hint(depth),
+            Schema::arbitrary_request_size_hint(depth),
+        ])
+    }
+}
+
+fn drop_some_entities(entities: Entities, u: &mut Unstructured<'_>) -> arbitrary::Result<Entities> {
+    let should_drop: bool = u.arbitrary()?;
+    if should_drop {
```
Comment (khieta) [inline cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs:105]: What's the rationale for this function?

File: cedar-drt/fuzz/src/lib.rs [hunk 10]
```diff
@@ -57,6 +60,22 @@ impl<'e> DifferentialTester<'e> {
         }
     }
 
+    /// Differentially test evaluating the given expression.
+    /// `r` and `entities` are used to populate the evaluator.
+    pub fn run_eval_test(&self, r: &ast::Request, expr: &Expr, entities: &Entities) -> bool {
+        let exts = Extensions::all_available();
+        let eval = match Evaluator::new(r, entities, &exts) {
```
Comment (khieta) [inline cedar-drt/fuzz/src/lib.rs:67]: Probably want the use of extensions restricted by a flag (like we have for `ABACSettings`)

**Review by khieta — APPROVED** [review]
Lgtm. Left a couple minor comments.

File: cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs (hunk 9 above)
Comment (aaronjeline) [reply cedar-drt/fuzz/fuzz_targets/eval-type-directed.rs:105]: Added in comments

**Review by shaobo-he-aws — APPROVED** [review]

**Pull request merged**